### ISSUE TITLE
Interaction with current-column

### ISSUE THREAD

**Issue by brownts** [jdtsmith/indent-bars#82]
I'm running into an issue when `indent-bars-mode` is enabled where it appears the `current-column` function may not report the correct value.  So far, I'm seeing this when using the treesitter region indentation implementation (i.e., `treesit-indent-region`) on a specific incorrectly indented section of code.  When I re-indent without `indent-bars-mode` enabled, this works fine and always indents correctly.  However, when `indent-bars-mode` is enabled, this same incorrectly indented section of code is never indented properly (so it's repeatable).

My hypothesis is that as `treesit-indent-region` walks through the sequential lines of the buffer, that `indent-bars-mode` is updating the scoping and somehow that is affecting the computation of `current-column`.  I had instrumented the `treesit-indent-region` function to determine that the call to `current-column` was reporting the incorrect column when visiting certain markers in the buffer.  Looking at the documentation for `current-column`, it indicates that the column is calculated from "the widths of all the display representations of the characters between the start of the previous line and point".  I'm wondering if the "visual annotation" (for lack of a better description) implemented by `indent-bars-mode` is somehow causing the column to be calculated incorrectly.

Can you think of any way this could be occurring?

I've attached a minimal configuration that can be used to reproduce this issue.  I've observed this issue on Windows and Linux and both Emacs 29 and Emacs 30, so I think it shouldn't be hard to reproduce.  I've also attached a test file (i.e., test.gpr) as a starting point.

[init.el.zip](https://github.com/user-attachments/files/17484802/init.el.zip)
[test.gpr.zip](https://github.com/user-attachments/files/17484808/test.gpr.zip)

- Scenario 1: `gpr-ts-mode` only, indentation works correctly
  - Open `test.gpr`, should activate `gpr-ts-mode` automatically.
  - Select entire buffer (`C-x h`)
  - Press `TAB` to activate region indentation
  - Diff buffer with file contents (i.e., `M-x diff-buffer-with-file RET RET`)
  - Observe that only a single line has been changed (indentation corrected)
  - ```diff
    --- /home/troy/junk/test.gpr
    +++ #<buffer test.gpr>
    @@ -6,7 +6,7 @@
             ("../docs/gtkada_ug/_build/html/*",
              "../docs/gtkada_ug/_build/latex/GtkAda.pdf");
 
    -case Blah is
    +      case Blah is
              when "YES" =>
                 null;
              when "NO" =>
    ```
- Scenario 2: `gpr-ts-mode` and `indent-bars-mode`, indentation doesn't work correctly
  - Revert previous changes in buffer (`C-x x g`)
  - Enable `indent-bars-mode` (`M-x indent-bars-mode`)
  - Select entire buffer (`C-x h`)
  - Press `TAB` to activate region indentation
  - Diff buffer with file contents (i.e., `M-x diff-buffer-with-file RET RET`)
  - Observe that while the original incorrectly indented line has been corrected, additional lines are now indented incorrectly.  These incorrect indentations are due to `current-column` reporting the wrong value when visiting the corresponding markers (which are placed just before "case Blah is" -- i.e., the "when" and "end case" are anchored to the start of the case statement).
  - ```diff
    --- /home/troy/junk/test.gpr
    +++ #<buffer test.gpr>
    @@ -6,12 +6,12 @@
             ("../docs/gtkada_ug/_build/html/*",
              "../docs/gtkada_ug/_build/latex/GtkAda.pdf");
 
    -case Blah is
    +      case Blah is
              when "YES" =>
                 null;
    -         when "NO" =>
    -            null;
    -      end case;
    +          when "NO" =>
    +             null;
    +       end case;
 
           for Artifacts ("share/doc/gtkada") use
             ("../docs/gtkada_rm/gtkada_rm");
    ``` 

It should also be noted, that even with `indent-bars-mode` enabled, I can manually move to each line of the incorrect indentation above and hit TAB to indent each line individually and it will correct the indentation on the lines.  Therefore, I think this issue may be timing related where the region indentation algorithm is running each of these indentations in quick succession.

I'm also including @casouri who implemented the treesitter support in Emacs and may have some insight as to what is happening in this situation.

**Comment by casouri** [comment]
I see, tree-sitter might need to use some other function that isn't affected by display property but can still calculates tab width correctly. Anyone knows if such function exists already?

**Comment by yantar92** [comment]
Yuan Fu ***@***.***> writes:

> I see, tree-sitter might need to use some other function that isn't affected by display property but can still calculates tab width correctly. Anyone knows if such function exists already?

See `org-current-text-column'

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by jdtsmith** [comment]
Not sure I fully follow this diagnosis.  `indent-bars` does not alter the text content in any way in the buffer, so should not be affecting current-column.  Do you have `indent-tabs-mode` enabled in this buffer?  In that case, it does apply a `display` property to tabs as needed, replacing them with the correct number of styled spaces as indicated in `tab-width`.   What if you disable `indent-tabs-mode` (and re-indent the entire buffer to replace with spaces)?  Does that resolve the issue?

**Comment by brownts** [comment]
> Not sure I fully follow this diagnosis. `indent-bars` does not alter the text content in any way in the buffer, so should not be affecting current-column. Do you have `indent-tabs-mode` enabled in this buffer? In that case, it does apply a `display` property to tabs as needed, replacing them with the correct number of styled spaces as indicated in `tab-width`. What if you disable `indent-tabs-mode` (and re-indent the entire buffer to replace with spaces)? Does that resolve the issue?

While in the particular configuration I provided above, I did not explicitly disable `indent-tabs-mode`, I do have it disabled in my personal configuration.  Regardless, you can add `M-x indent-tabs-mode` to the above instructions to disable that mode in the above and it doesn't affect the outcome.  The indentation is still indented incorrectly when `indent-bars-mode` is enabled.  Additionally, there are no literal TAB characters in the provided example file.  I don't believe this has anything to do with the actual text in the buffer.

The region indentation algorithm (i.e., `treesit-indent-region`) for the above case statement creates 3 separate markers (all positioned directly before "case Blah is" in the buffer), to anchor the two separate "when" lines and the "end case".  When visiting the anchor from the line of the first "when", `current-column` is reported correctly (i.e., 6), but when visited from the line of the second 'when' and the 'end case' line, `current-column` reports 7, all visiting markers that point at the same buffer position, but from different lines.  I verified that all 3 markers indicate the same buffer position.  That's why I suspected `current-column` to be the problem.

> See `org-current-text-column`

Thanks for the tip @yantar92.  I took a look at the implementation of `org-current-text-column` and it essentially replaces `current-column` with:
```elisp
(string-width (buffer-substring-no-properties
                    (line-beginning-position) (point)))
```
I'm adding [this link](https://lists.gnu.org/archive/html/emacs-orgmode/2022-07/msg00462.html) for additional context on why `org-current-text-column` was created.

I then tried an experiment with `treesit-indent-region` where I replaced the call to `current-column` with the equivalent described above.

FROM:
```elisp
(save-excursion
  (goto-char anchor)
  (+ offset (current-column)))
```

TO:
```elisp
(save-excursion
  (goto-char anchor)
  (+ offset 
     (string-width (buffer-substring-no-properties
       (line-beginning-position) (point)))))
```

Repeating Scenario 2 above now no longer results in incorrect indentation.  Therefore, I think there is some sort of property introduced by `indent-bars-mode` that must be affecting the `current-column` calculation.

**Comment by yantar92** [comment]
Troy Brown ***@***.***> writes:

> I'm adding [this link](https://lists.gnu.org/archive/html/emacs-orgmode/2022-07/msg00462.html) for additional context on why `org-current-text-column` was created.

Also, see https://debbugs.gnu.org/cgi/bugreport.cgi?bug=56837

`current-column' and `current-indentation' are sensitive to how the text
is displayed and should be used with care when the indentation is a part
of syntax rather than simply a visual.

-- 
Ihor Radchenko // yantar92,
Org mode contributor,
Learn more about Org mode at <https://orgmode.org/>.
Support Org development at <https://liberapay.com/org-mode>,
or support my work at <https://liberapay.com/yantar92>

**Comment by brownts** [comment]
As a follow-on from the previous investigation, I was curious what the properties inserted by `indent-bars` looked like.  Therefore, I printed out the properties for the entire first line of the case statement from within `treesit-indent-region`.  As can be seen below, `indent-bars-display` is applied to the first 6 characters of the line, but it's string representation contains 7 spaces and the newline, thus it's not the same size as the range it's applied to.  I'm guessing this is why `current-column` reports a value of 7.

```text
#("      case Blah is" 0 6
  (indent-bars-display
   #("       \n" 3 4 (face indent-bars-1) 6 7 (face indent-bars-2))
   fontified nil)
  6 10 (face font-lock-keyword-face fontified nil) 10 11
  (fontified nil) 11 15
  (face font-lock-variable-use-face fontified nil) 15 16
  (fontified nil) 16 18 (face font-lock-keyword-face fontified nil))
```

**Comment by jdtsmith** [comment]
Hmm, definitely not good.  `indent-bars` should not be changing the width in any way.

# Verifying 

I'm a bit skeptical though.    The `indent-bars-display` property (an alias for `display`) you reference should be applied _only_ to a single newline on a blank line.  If you go to the beginning of the `case` line and `C-u C-x =` can you confirm that same properties you mention are found?  At a blank line, here's what `M-: (buffer-substring (line-beginning-position) (1+ (line-end-position)))` gives me (by example):

```
#("\n" 0 1 (indent-bars-display #("     \n" 4 5 (face indent-bars-1)) 
              face font-lock-doc-face indent-bars-font-lock-pending nil fontified t))
```

I.e. a single newline replaced with spaces and a newline.  This obviously would not impact column calculation, as the line is blank.

# To Try

If you validate the above, here are some things to try.  Note that I cannot test your reproducer because I don't have those TS backends (you could try replicating in python). 

- If you disable `indent-bars-display-on-blank-lines` and reset, does the problem go away?   If so, that would indicate blank line properties are "leaking" onto the next line for some reason.
- On that theory, I've pushed a commit which adds `rear-nonsticky t` to the properties added on newlines.  This should not be needed, but perhaps something treesitter is doing triggers inheritance of properties (inhibiting font-lock while altering whitespace?).
- If all else fails, can you please work up a simpler reproducer demonstrating this type of  "`indent-bars-display` with newline applied to on non-blank line" issue?   Please load _ONLY_ indent-bars (I see you have whitespace-mode enabled in your original recipe).  If possible, please reproduce in a python buffer.

**Comment by jdtsmith** [comment]
@yantar92 I'd be interested in hearing what issues you think might appear in `indent-bars`' due to its use of `current-column`, e.g. in Python buffers with outline-based invisibility.  I actually use this combo all the time and haven't found issues.  Is the problem if portions of lines are wrapped (vs. full lines)?

**Comment by brownts** [comment]
> If you validate the above, here are some things to try. Note that I cannot test your reproducer because I don't have those TS backends (you could try replicating in python).
> 

Not sure why you can't use the init.el I provided, as it points at the GitHub repos and would automatically install everything you need...oh well, it doesn't matter I'll provided a Python reproducer below, now that I have more insight into how to reproduce this.

> * If you disable `indent-bars-display-on-blank-lines` and reset, does the problem go away?   If so, that would indicate blank line properties are "leaking" onto the next line for some reason.

Yes, it appears so, from my limited testing.  I think you are on the right track here.

> * On that theory, I've pushed a commit which adds `rear-nonsticky t` to the properties added on newlines.  This should not be needed, but perhaps something treesitter is doing triggers inheritance of properties (inhibiting font-lock while altering whitespace?).

The example below doesn't use treesitter, but just uses regular `python-mode`.  I haven't tried this change yet, but hopefully that fixes it.

> * If all else fails, can you please work up a simpler reproducer demonstrating this type of  "`indent-bars-display` with newline applied to on non-blank line" issue?   Please load _ONLY_ indent-bars (I see you have whitespace-mode enabled in your original recipe).  If possible, please reproduce in a python buffer.

I don't believe `whitespace-mode` is enabled, maybe you're confusing that with a different issue?

Here is a Python example you can use to reproduce this:

```python
def do_something:
   def do_1_1:
      print "Hi"

def do_2_1:
         def do_3_1:
            print "Hi"
         def do_3_2:
            print "Hi"

         print "Hi"

      def do_2_2:
         print "Hi"

# Local Variables:
# mode: python
# python-indent-offset: 3
# indent-tabs-mode: nil
# End:
```

The above can be used to reproduce this issue.  First evaluate the function below in your scratch buffer, then place point in the buffer above on the "def do_2_1:" in the very first column of the line, turn on `indent-bars-mode` (`M-x indent-bars-mode`), and then issue: `M-: (test-column)`.  This essentially simulates what was happening in `treesit-indent-region`.  First it sets a marker on point, next it indents the line to column 6.  Then it moves forward 1 line, jumps to the previous marker to get the column and adds 3 to it to indent the "def do_3_1:" line.  This is repeated for the "def do_3_2:" line after that.  I also record point and column at the marker as well as the line including properties at that point in time and print them out at the end.

```elisp
(defun test-column ()
  (let ((anchor (copy-marker (point-marker) t))
	(p1 nil)
	(p2 nil)
	(c1 nil)
	(c2 nil)
	(l1 nil)
	(l2 nil))
    (indent-line-to 6)
    (forward-line 1)
    (indent-line-to
     (save-excursion
       (goto-char anchor)
       (setq p1 (point))
       (setq l1 (buffer-substring (line-beginning-position) (line-end-position)))
       (setq c1 (current-column))
       (+ 3 c1)))
    (forward-line 2)
    (indent-line-to
     (save-excursion
       (goto-char anchor)
       (setq p2 (point))
       (setq l2 (buffer-substring (line-beginning-position) (line-end-position)))
       (setq c2 (current-column))
       (+ 3 c2)))
    (message "L1 (point=%d, column=%d):" p1 c1)
    (pp l1)
    (message "L2 (point=%d, column=%d): " p2 c2)
    (pp l2)
    (message "DONE")))
```

When I run this, I see that "def do_3_2:" is indented incorrectly, and the following is output in the Messages buffer.  Note that the column output in the second line (i.e., "L2") is incorrect.  The properties from the line do seem to include the content from the blank line prior to "def do_2_1:", so your suspicion about the bleed is likely correct.

```text
L1 (point=58, column=6):
#("      def do_2_1:" 0 6
  (indent-bars-display #("    \n" 3 4 (face indent-bars-1)) fontified
		       nil)
  6 9 (face font-lock-keyword-face fontified nil) 9 10 (fontified nil)
  10 16 (face font-lock-function-name-face fontified nil) 16 17
  (fontified nil))

L2 (point=58, column=4): 
#("      def do_2_1:" 0 6
  (indent-bars-display #("    \n" 3 4 (face indent-bars-1)) fontified
		       nil)
  6 9 (face font-lock-keyword-face fontified nil) 9 10 (fontified nil)
  10 16 (face font-lock-function-name-face fontified nil) 16 17
  (fontified nil))

DONE
"DONE"
```

**Comment by jdtsmith** [comment]
Thanks for the test and command.  I've run it as directed and do not encounter property bleed, with the release version or current head:

```
Indent-Bars mode enabled in current buffer
L1 (point=69, column=6):
#("      def do_2_1:" 0 6
  (fontified nil)
  6 9
  (face font-lock-keyword-face fontified nil)
  9 10
  (fontified nil)
  10 16
  (face font-lock-function-name-face fontified nil)
  16 17
  (fontified nil))

L2 (point=69, column=6): 
#("      def do_2_1:" 0 6
  (fontified nil)
  6 9
  (face font-lock-keyword-face fontified nil)
  9 10
  (fontified nil)
  10 16
  (face font-lock-function-name-face fontified nil)
  16 17
  (fontified nil))
```

These are as expected.  The main point to consider is that `indent-line-to` will alter the text, which will cause the affected line(s) to be refontified.  You are pulling text from the buffer in the same redisplay cycle, likely prior to font-lock having a chance to update.  

This is why earlier I asked "If you go to the beginning of the case line and C-u C-x = can you confirm that same properties you mention are found?".  I.e. after redisplay and font-lock happen.  Can you check this and report?

You may as well go ahead and try my "guess" at a fix on main.  Can you also mention your version?  You are running from emacs -q with no other packages enabled?  One other question: if you disable treesitter support, restart indent-bars, and run your simple test above, does the issue persist?

**Comment by brownts** [comment]
TLDR; Using your latest "rear-nonsticky" commit, seems to have fixed this.

> These are as expected.

So you weren't able to reproduce the issue?  Hmm...very frustrating.

> This is why earlier I asked "If you go to the beginning of the case line and C-u C-x = can you confirm that same properties you mention are found?". I.e. after redisplay and font-lock happen. Can you check this and report?

Running `C-u C-x =` does not show this afterwards, therefore, it's transitional.  I speculated in my original post that it may be timing related.  I've seen it fail on the first visit to the marker (rather than the second visit) if I was printing out information during the excursion.  Since it's possible there's a timing aspect to it, you might get it to fail by extending/modifying the test.

> Can you also mention your version? You are running from emacs -q with no other packages enabled? One other question: if you disable treesitter support, restart indent-bars, and run your simple test above, does the issue persist?

I was using the latest in the repository at the time I started testing this, which was the v0.8.1 tag: d5dfdea4f9e5dad8e93fd7a0cf86e6a1a8776bde

Since I had provided a minimal `init.el` in the original post, I was using that configuration, so that we'd both be running with the same setup.  I just created a temporary directory (i.e., `~/.emacs-indent-region.d`), placed the `init.el` in there, and started Emacs with the `--init-directory` option, specifying that directory.  I guess you could run with `-q` if you add your package to the load path.  The `init.el` uses the ":vc" keyword to pull from GitHub which isn't available in Emacs 29, so you can comment that out after first running it with Emacs 30 to pull from GitHub, although you can install it all manually as well if you like.  I've run this test on both Emacs 29 and Emacs 30 and observed the same failure using the commit mentioned above.

> You may as well go ahead and try my "guess" at a fix on main.

Using your latest commit, running both my original test as well as the Python test now work on Emacs 29 and Emacs 30.  So, I would say that with my limited testing, this seems to have fixed the issue.
 
```text
Indent-Bars mode enabled in current buffer
Mark set
test-column
L1 (point=58, column=6):
#("      def do_2_1:" 0 6 (fontified nil) 6 9
  (face font-lock-keyword-face fontified nil) 9 10 (fontified nil) 10
  16 (face font-lock-function-name-face fontified nil) 16 17
  (fontified nil))

L2 (point=58, column=6): 
#("      def do_2_1:" 0 6 (fontified nil) 6 9
  (face font-lock-keyword-face fontified nil) 9 10 (fontified nil) 10
  16 (face font-lock-function-name-face fontified nil) 16 17
  (fontified nil))

DONE
"DONE"
```

**Comment by jdtsmith** [comment]
> Running C-u C-x = does not show this afterwards, therefore, it's transitional.
> Using your latest commit, running both my original test as well as the Python test now work on Emacs 29 and Emacs 30.

Thanks for confirming.  So it seems my defensive measure to prevent even transitory sticky inheritance when adding spaces after a filled blank-line was the right approach.  Given how complex the buffer edits->redisplay->change hooks->jit-lock->redisplay cycles are, it doesn't surprise me that we get different results here.   There are other issues related to adding `display` to the `\n` on a line, but unfortunately there's nowhere else to put them (a fully blank line is _only_ an `\n`).

**Comment by yantar92** [comment]
> @yantar92 I'd be interested in hearing what issues you think might appear in `indent-bars`' due to its use of `current-column`, e.g. in Python buffers with outline-based invisibility. I actually use this combo all the time and haven't found issues. Is the problem if portions of lines are wrapped (vs. full lines)?

Maybe https://debbugs.gnu.org/cgi/bugreport.cgi?bug=64696 can give some insights
